File: src/motor_ai_sim/simulation/eddy_steps.py
"""Default time resolution of an eddy-current (sigma*dA/dt) run.

Owner decision 2026-09-26: a run with the coupled eddy solve ON defaults to 72
steps per electrical period, not 36.  Measured
(``docs/EDDY_TIME_INTEGRATION_2026-09-25.md`` section 1, L155 rated, BDF2 with the
midpoint loss): the magnet loss reads -4.3 % against the step-converged limit at
36 steps and -1.05 % at 72; shaft -3.1 / -0.7 %, copper AC -1.1 / -0.3 %.

It is a DEFAULT and nothing more: a step count the caller names is always the
one solved (the slip-ring snap in the solver may still move it to a divisor of
the ring, and the result says so in ``n_steps_per_period_requested`` /
``steps_snapped``).  Optimizer screening does not use this number — its
candidates carry ``sampling_purpose="optimization"`` and their own count.
"""
from __future__ import annotations

from typing import Any, Optional

#: steps per electrical period of an eddy-on run whose request names none
EDDY_DEFAULT_STEPS_PER_PERIOD = 72
#: the transient route's default for an eddy-OFF run (unchanged)
STATIC_DEFAULT_STEPS_PER_PERIOD = 60


def validate_steps_per_period(value: Any, *, name: str = "n_steps_per_period") -> int:
    """A named step count, checked: a whole number >= 1, never a bool."""
    if isinstance(value, bool):
        raise ValueError("%s must be a whole number of steps; got %r" % (name, value))
    try:
        f = float(value)
    except (TypeError, ValueError):
        raise ValueError("%s must be a whole number of steps; got %r" % (name, value))
    if f != f or f in (float("inf"), float("-inf")) or f != int(f):
        raise ValueError("%s must be a whole number of steps; got %r" % (name, value))
    if int(f) < 1:
        raise ValueError("%s must be >= 1; got %r" % (name, value))
    return int(f)


def resolve_steps_per_period(value: Optional[Any], *, eddy: bool) -> int:
    """The step count a transient solves: the caller's, else the default.

    ``None`` (or ``""``) = not named -> 72 with the eddy solve on, 60 without.
    Anything named is validated and returned unchanged.
    """
    if value is None or (isinstance(value, str) and not value.strip()):
        return (EDDY_DEFAULT_STEPS_PER_PERIOD if bool(eddy)
                else STATIC_DEFAULT_STEPS_PER_PERIOD)
    return validate_steps_per_period(value)

File: web/src/lib/eddySteps.ts
/**
 * Steps per electrical period — the eddy-run default and the slip-ring snap.
 *
 * Owner decision 2026-09-26: runs with the coupled eddy solve (the Simulation
 * tab — eddy is always on there — the coupled loop, and the duty runs a report
 * quotes) default to 72 steps per period instead of 36/40.  BDF2 reads the L155
 * magnet loss −4.3 % at 36 steps and −1 % at 72
 * (docs/EDDY_TIME_INTEGRATION_2026-09-25.md §1).  The backend uses the same
 * number when a request names none (simulation/eddy_steps.py).
 *
 * It is a DEFAULT: the picker stays visible and any count the user picks is the
 * one solved («должен быть всегда выбор»).  Optimizer screening keeps its own
 * count while the tab holds only this default (routes/optimization.py).
 *
 * Pure functions only — `lib/__tests__/eddySteps.test.mjs` restates them.
 */

export const EDDY_DEFAULT_STEPS = 72;

/** The tab's factory defaults before 2026-09-26.  A count equal to one of them
 *  with no record of who set it is the old default, not a choice. */
export const LEGACY_DEFAULT_STEPS: readonly number[] = [36, 40];

/** Who set the count: the eddy default, or the user. */
export type StepsSource = 'eddy_default' | 'user';

export const isStepsSource = (v: unknown): v is StepsSource =>
  v === 'eddy_default' || v === 'user';

/** Nearest count the sliding band can run: a divisor of the ring's nodes per
 *  period (ties go up).  Above the ring any count is legal — the solver raises
 *  the ring to it — so it is returned unchanged.  Same rule as the solver's
 *  `_snap_steps_to_nodes` and the panel's picker. */
export function snapStepsToRing(v: number, ring: number): number {
  if (!(ring > 0) || !(v > 0)) return v;
  if (v > ring) return v;
  if (ring % v === 0) return v;
  let best = ring;
  for (let d = 1; d <= ring; d++)
    if (ring % d === 0 && (Math.abs(d - v) < Math.abs(best - v)
        || (Math.abs(d - v) === Math.abs(best - v) && d > best))) best = d;
  return best;
}

/** What the panel adopts from the shared config on mount.
 *
 *  `stored` is the config's count (else this browser's), `source` the config's
 *  record of who set it.  A recorded source is obeyed as is.  With no record:
 *  nothing stored, or one of the old factory defaults → the eddy default
 *  (`migratedFrom` names the old count so the panel can say so); any other
 *  count was chosen by someone and is kept as the user's. */
export function adoptSteps(stored: number | null, source: unknown): {
  steps: number; source: StepsSource; migratedFrom: number | null;
} {
  const n = typeof stored === 'number' && Number.isFinite(stored) && stored > 0
    ? Math.round(stored) : null;
  if (source === 'user' && n !== null)
    return { steps: n, source: 'user', migratedFrom: null };
  if (source === 'eddy_default')
    return { steps: n ?? EDDY_DEFAULT_STEPS, source: 'eddy_default', migratedFrom: null };
  if (n === null)
    return { steps: EDDY_DEFAULT_STEPS, source: 'eddy_default', migratedFrom: null };
  if (LEGACY_DEFAULT_STEPS.includes(n) && n !== EDDY_DEFAULT_STEPS)
    return { steps: EDDY_DEFAULT_STEPS, source: 'eddy_default', migratedFrom: n };
  return { steps: n, source: 'user', migratedFrom: null };
}

/** ONE line, or null, when the count that runs is not the count asked for. */
export function stepsNote(requested: number, ran: number, ring: number,
                          migratedFrom: number | null = null): string | null {
  if (migratedFrom !== null) {
    return `Steps ${migratedFrom} → ${ran}: eddy runs now default to ${EDDY_DEFAULT_STEPS}`
      + (ran !== EDDY_DEFAULT_STEPS
        ? ` (${ran} = nearest divisor of the ${ring}-node slip ring)` : '')
      + '.';
  }
  if (requested === ran) return null;
  return `Steps ${requested} → ${ran}: the count must divide this machine's `
    + `${ring}-node slip ring.`;
}
